File: lib/alphaVantage.ts
import "server-only";

const BASE = "https://www.alphavantage.co/query";

const memoryCache = new Map<string, { value: unknown; expires: number }>();

function getCache<T>(key: string): T | undefined {
  const entry = memoryCache.get(key);
  if (!entry) return undefined;
  if (entry.expires < Date.now()) {
    memoryCache.delete(key);
    return undefined;
  }
  return entry.value as T;
}

function setCache(key: string, value: unknown, ttlMs: number) {
  memoryCache.set(key, { value, expires: Date.now() + ttlMs });
}

function getApiKey() {
  const key = process.env.ALPHA_VANTAGE_API_KEY;
  if (!key) throw new Error("Missing ALPHA_VANTAGE_API_KEY env var.");
  return key;
}

async function avFetch<T>(params: Record<string, string>, ttlMs: number) {
  const key = `av:${JSON.stringify(params)}`;
  const cached = getCache<T>(key);
  if (cached) return cached;

  const search = new URLSearchParams({ ...params, apikey: getApiKey() });
  const url = `${BASE}?${search.toString()}`;
  const res = await fetch(url, { cache: "no-store" });
  if (!res.ok) throw new Error(`Alpha Vantage HTTP ${res.status}`);
  const json = (await res.json()) as Record<string, unknown> & T;

  if ((json as { Note?: string }).Note) {
    throw new Error("Alpha Vantage rate limit reached. Try again in a minute.");
  }
  if ((json as { Information?: string }).Information) {
    throw new Error(
      `Alpha Vantage info: ${(json as { Information?: string }).Information}`,
    );
  }
  if ((json as { "Error Message"?: string })["Error Message"]) {
    throw new Error(
      `Alpha Vantage error: ${
        (json as { "Error Message"?: string })["Error Message"]
      }`,
    );
  }

  setCache(key, json, ttlMs);
  return json;
}

export type GlobalQuote = {
  ticker: string;
  price: number;
  changePercent: number;
  asOf: string;
};

export async function getGlobalQuote(symbol: string): Promise<GlobalQuote> {
  const json = await avFetch<{ "Global Quote"?: Record<string, string> }>(
    { function: "GLOBAL_QUOTE", symbol },
    60_000,
  );
  const quote = json["Global Quote"];
  if (!quote || !quote["05. price"]) {
    throw new Error(`No quote returned for ${symbol}.`);
  }
  return {
    ticker: symbol,
    price: Number(quote["05. price"]),
    changePercent: Number(
      (quote["10. change percent"] || "0").replace("%", ""),
    ),
    asOf:
      quote["07. latest trading day"] ?? new Date().toISOString().slice(0, 10),
  };
}

export type MonthlySeries = {
  ticker: string;
  prices: { date: string; close: number }[];
};

export async function getMonthlyAdjusted(symbol: string): Promise<MonthlySeries> {
  const json = await avFetch<{
    "Monthly Adjusted Time Series"?: Record<string, Record<string, string>>;
  }>(
    { function: "TIME_SERIES_MONTHLY_ADJUSTED", symbol },
    24 * 60 * 60_000,
  );

  const series = json["Monthly Adjusted Time Series"];
  if (!series) throw new Error(`No monthly series for ${symbol}.`);

  const prices = Object.entries(series)
    .map(([date, fields]) => ({
      date,
      close: Number(fields["5. adjusted close"] ?? fields["4. close"]),
    }))
    .filter((row) => Number.isFinite(row.close))
    .sort((a, b) => (a.date < b.date ? -1 : 1));

  return { ticker: symbol, prices };
}

export function sliceWindow(series: MonthlySeries, start: string, end: string) {
  return {
    ticker: series.ticker,
    prices: series.prices.filter((row) => row.date >= start && row.date <= end),
  } satisfies MonthlySeries;
}

export function monthlyReturns(series: MonthlySeries): number[] {
  const out: number[] = [];
  for (let i = 1; i < series.prices.length; i++) {
    const prev = series.prices[i - 1].close;
    const curr = series.prices[i].close;
    if (prev > 0) out.push((curr - prev) / prev);
  }
  return out;
}

File: ml/alpha_vantage.py
"""Alpha Vantage data fetcher with caching."""

import os
import time
import requests
import pandas as pd
from functools import lru_cache

BASE = "https://www.alphavantage.co/query"

_cache: dict[str, tuple[float, pd.DataFrame]] = {}
CACHE_TTL = 86400  # 24 hours
_last_request_time = 0.0
REQUEST_DELAY = 1.5  # seconds between API calls (free tier: 1/sec)


def _api_key() -> str:
    key = os.getenv("ALPHA_VANTAGE_API_KEY", "")
    if not key:
        raise ValueError("ALPHA_VANTAGE_API_KEY not set")
    return key


def _fetch(params: dict) -> dict:
    global _last_request_time
    now = time.time()
    wait = REQUEST_DELAY - (now - _last_request_time)
    if wait > 0:
        time.sleep(wait)

    params["apikey"] = _api_key()
    r = requests.get(BASE, params=params, timeout=30)
    _last_request_time = time.time()
    r.raise_for_status()
    data = r.json()
    if "Note" in data or "Information" in data:
        raise RuntimeError(data.get("Note") or data.get("Information", "Rate limited"))
    if "Error Message" in data:
        raise RuntimeError(data["Error Message"])
    return data


def get_monthly_adjusted(symbol: str) -> pd.DataFrame:
    """Return monthly adjusted close prices for a symbol."""
    cache_key = f"monthly:{symbol}"
    if cache_key in _cache:
        ts, df = _cache[cache_key]
        if time.time() - ts < CACHE_TTL:
            return df

    data = _fetch({"function": "TIME_SERIES_MONTHLY_ADJUSTED", "symbol": symbol})
    series = data.get("Monthly Adjusted Time Series", {})
    if not series:
        raise RuntimeError(f"No monthly data for {symbol}")

    rows = []
    for date_str, fields in series.items():
        rows.append({
            "date": pd.Timestamp(date_str),
            "open": float(fields.get("1. open", 0)),
            "high": float(fields.get("2. high", 0)),
            "low": float(fields.get("3. low", 0)),
            "close": float(fields.get("4. close", 0)),
            "adj_close": float(fields.get("5. adjusted close", fields.get("4. close", 0))),
            "volume": float(fields.get("6. volume", 0)),
            "dividend": float(fields.get("7. dividend amount", 0)),
        })

    df = pd.DataFrame(rows).sort_values("date").reset_index(drop=True)
    _cache[cache_key] = (time.time(), df)
    return df


def get_daily(symbol: str, outputsize: str = "compact") -> pd.DataFrame:
    """Return daily close prices (last 100 trading days by default)."""
    cache_key = f"daily:{symbol}:{outputsize}"
    if cache_key in _cache:
        ts, df = _cache[cache_key]
        if time.time() - ts < CACHE_TTL:
            return df

    data = _fetch({
        "function": "TIME_SERIES_DAILY_ADJUSTED",
        "symbol": symbol,
        "outputsize": outputsize,
    })
    series = data.get("Time Series (Daily)", {})
    if not series:
        raise RuntimeError(f"No daily data for {symbol}")

    rows = []
    for date_str, fields in series.items():
        rows.append({
            "date": pd.Timestamp(date_str),
            "open": float(fields.get("1. open", 0)),
            "high": float(fields.get("2. high", 0)),
            "low": float(fields.get("3. low", 0)),
            "close": float(fields.get("4. close", 0)),
            "adj_close": float(fields.get("5. adjusted close", fields.get("4. close", 0))),
            "volume": float(fields.get("6. volume", 0)),
        })

    df = pd.DataFrame(rows).sort_values("date").reset_index(drop=True)
    _cache[cache_key] = (time.time(), df)
    return df


def get_global_quote(symbol: str) -> dict:
    """Return current price info for a symbol."""
    data = _fetch({"function": "GLOBAL_QUOTE", "symbol": symbol})
    quote = data.get("Global Quote", {})
    return {
        "ticker": symbol,
        "price": float(quote.get("05. price", 0)),
        "change_pct": float(quote.get("10. change percent", "0").replace("%", "")),
    }
